api/src/main/java/io/smallrye/stork/api/Service.java

```diff
@@ -12,39 +15,110 @@
  */
 public class Service {
 
+    private final Lock instanceSelectionLock;
     private final Optional<LoadBalancer> loadBalancer;
     private final ServiceDiscovery serviceDiscovery;
     private final String serviceName;
     private final boolean secure;
 
-    public Service(String serviceName, Optional<LoadBalancer> loadBalancer, ServiceDiscovery serviceDiscovery, boolean secure) {
+    public Service(String serviceName, Optional<LoadBalancer> loadBalancer, ServiceDiscovery serviceDiscovery,
+            boolean secure, boolean threadSafeLoadBalancer) {
         this.loadBalancer = loadBalancer;
         this.serviceDiscovery = serviceDiscovery;
         this.serviceName = serviceName;
         this.secure = secure;
+        instanceSelectionLock = threadSafeLoadBalancer ? null : new ReentrantLock();
```

Comment:
why reentrant?

Reply:
That's the simplest thing java brings...
I think I'll change it to semaphore. The reentrant lock is bound to a thread, a single thread can enter it multiple times. I don't think we want it.